src/core/cgroup.c

```diff
@@ -1803,10 +1815,87 @@ static void unit_remove_from_cgroup_realize_queue(Unit *u) {
         u->in_cgroup_realize_queue = false;
 }
 
+/* Controllers can only be enabled breadth-first, from the root of the
+ * hierarchy downwards to the unit in question. */
+static int unit_realize_cgroup_now_enable_breadth_first(Unit *u, ManagerState state) {
```

Comment:
Hmm, I'm confused by "breadth" here. The description says that enabling must be done "downwards", and disabling "upwards". In fact you say that "Notice that we need to touch j in one direction, but not the other.", i.e. there's no breadth, just a straight parents-first iteration.